Desktop: Add `enex` to `html` export

devonzuegel opened the pull request:
## Features introduced in this PR

Adds a new <kbd>ENEX – Evernote Export File (as HTML)</kbd> importer (as discussed [in the forum](https://discourse.joplinapp.org/t/evernote-import-discards-styles/113/12)):

<img width="55%" src="https://user-images.githubusercontent.com/6979755/62836020-0eef7580-bc5f-11e9-9b1f-fc9d8aa8cf93.png" />

It takes the raw XML in the `.enex` file (which is very close to HTML), replaces the `en-*` tags with the appropriate HTML elements, and imports the resulting HTML as notes.

## Refactored in this PR

- Extracted `lib/resourceUtils.js` for shared utility functions used in both `lib/import-enex-md-gen.js` and `lib/import-enex-html-gen.js`: [github.com/laurent22/joplin/pull/1795/files#diff-e16a7261ac899d30435dab0bf1b3dce9](https://github.com/laurent22/joplin/pull/1795/files#diff-e16a7261ac899d30435dab0bf1b3dce9)

- Previously, the `InteropService` for importing only looked at the `format` of the file to determine which importer to use. Now, we have 2 importer options for a single file format (`enex`), so I added an `importerClass` attribute so you can paying attention to which importer was chosen by the user: [github.com/laurent22/joplin/pull/1795/files#diff-a65832a1338a095dea8f4d4ff673db77R82](https://github.com/laurent22/joplin/pull/1795/files#diff-a65832a1338a095dea8f4d4ff673db77R82). This entails passing in more of the options in here: [github.com/laurent22/joplin/pull/1795/files#diff-a65832a1338a095dea8f4d4ff673db77R197](https://github.com/laurent22/joplin/pull/1795/files#diff-a65832a1338a095dea8f4d4ff673db77R197)

## Questions
I'd love your thoughts on the following:
- Would you like me to split this out into 2-3 PRs, so that it's easier to review? I could make 1 for the refactoring, 1 for the backend importer, and 1 for the UI. I realize that this PR is a bit large.
- Want me to squash the commits? They're a bit messy.
- Want me to add tests for the UI or any additional cases?
- Instead of relying on the `clean-html` module, which adds a new dependency, would you prefer I implement something more lean like [this](https://github.com/isaacs/sax-js/blob/5aee2163d55cff24b817bbf550bac44841f9df45/examples/pretty-print.js) to obviate that additional dependency?
- When the HTML importer sees an `en-todo`, it currently outputs `'<input type="checkbox" onclick="return false;" />'`. This is readonly in a preview pane (i.e. you have to go into the HTML in the editor to toggle a checkbox), which is not ideal since the checkboxes in the Markdown importer are toggle-able from the preview pane. That said, I don't think it's a huge problem because (a) you're most likely to be doing an import of old notes whose state you don't want to change much but rather treat more as an archive and (b) you're most likely to want the HTML importer for web clippings, which generally don't have checkboxes in the first place.
    
    With that background, my question is this: would you like me to make imported checkboxes clickable?

## Next steps

I'd love to also add an importer that uses the HTML importer if it's a clipping and converts to Markdown if it's a user-made note. I see two ways of doing this:

1. We could add another importer to the existing list, something like: <img width="60%" src="https://user-images.githubusercontent.com/6979755/62836239-81615500-bc61-11e9-98b6-0ed352f99000.png" />
2. We could consolidate the 3 `.enex` importers to a single `ENEX - Evernote Export File` option, and when that's selected have a pop-up where the user specifies whether they want clipped and personal notes imported as Markdown or HTML.

Option 2 seems more elegant to me, plus it gives us space to offer a bit more information about the consequences of the options, but it does deviate from the current import model more.

devonzuegel commented on ReactNativeClient/lib/import-enex-html-gen.js:
```diff
@@ -0,0 +1,184 @@
+		 */
+		lines.push(resourceUtils.audioElement({src, id: resource.id}));
+	} else {
+		// TODO: figure out if we need to handle other mime types
```
Are there other image mime types that should be handled here?

devonzuegel commented on ReactNativeClient/lib/resourceUtils.js:
```diff
@@ -0,0 +1,84 @@
+		`    ${alt || src || id || 'Download audio'}`,
+		'  </a>',
+		'</p>',
+	].join('');
```
This creates an audio player + a link to download the element, and the rendered output looks like this:
![image](https://user-images.githubusercontent.com/6979755/62837064-29c7e700-bc6b-11e9-9781-43c3cee02f59.png)

laurent22 commented:
Thanks for the pull request @devonzuegel and sorry for the delay in replying. I was in vacation in August and didn't have time to really look into the code till now.

> Would you like me to split this out into 2-3 PRs, so that it's easier to review? I could make 1 for the refactoring, 1 for the backend importer, and 1 for the UI. I realize that this PR is a bit large.

It's fine as it is as the feature is quite self contained.

> Want me to squash the commits? They're a bit messy.

We always squash and merge so multiple commits are fine.

> Want me to add tests for the UI or any additional cases?

The tests you've added are fine I think. As for the UI, we currently don't have any way to test it unfortunately.

> Instead of relying on the clean-html module, which adds a new dependency, would you prefer I implement something more lean like this to obviate that additional dependency?

I don't quite understand the need for HTML cleaning. The Evernote HTML should already be valid, shouldn't it?

> When the HTML importer sees an en-todo, it currently outputs '<input type="checkbox" onclick="return false;" />'. This is readonly in a preview pane (i.e. you have to go into the HTML in the editor to toggle a checkbox), which is not ideal since the checkboxes in the Markdown importer are toggle-able from the preview pane. That said, I don't think it's a huge problem because (a) you're most likely to be doing an import of old notes whose state you don't want to change much but rather treat more as an archive and (b) you're most likely to want the HTML importer for web clippings, which generally don't have checkboxes in the first place.

I agree with that, we can leave it as you did.

laurent22 commented on ReactNativeClient/lib/services/InteropService.js:
```diff
@@ -130,6 +139,15 @@ class InteropService {
 		return output;
 	}
 
+	newModuleFromPath_(options) {
```
Any reason why this function was needed to load the modules?

devonzuegel replied:
The existing `newModule_` fn would load by the input format. This was fine when there was a 1-1 mapping of input formats to output formats, but now that we have 2 possible outputs for an `enex` input, we need to be explicit with which importer we want to use.

It might make sense to simply move all the existing formatters to the `newModuleFromPath` approach, so that there's only one way to do this mapping. Would you like me to do that? If so, I'd prefer to do that refactoring in a separate PR since this one is already quite large.

devonzuegel replied:
Btw I pushed a commit to update the existing function names to make this slightly more clear in the meantime.

laurent22 replied:
Ok that makes sense. No need to refactor for now, it can be done if it ever becomes necessary. If you could add the explanation above as a comment for now that would be great.

laurent22 commented on ReactNativeClient/lib/resourceUtils.js:
```diff
@@ -0,0 +1,84 @@
+
+const attachmentElement = ({src, attributes, id}) =>
+	[
+		`<a href='#' ${attributesToStr(attributes)} ${ipcProxySendToHost(id)}>`,
```
What does this code do? Is it to load audio files?

devonzuegel replied:
Given the source for a file whose type is otherwise not supported, it's a helper to create an `<a /`> link to the file in the HTML.

laurent22 commented:
Otherwise I've checked the code and it's pretty much ready to merge. If you could just check the few comments I left that would be great.

devonzuegel commented:
No worries on the delay! Thank you for the review, and I hope you had a wonderful vacation.

> > Instead of relying on the clean-html module, which adds a new dependency, would you prefer I implement something more lean like this to obviate that additional dependency?
> 
> I don't quite understand the need for HTML cleaning. The Evernote HTML should already be valid, shouldn't it?

Yeah, Evernote HTML is valid, but (1) the default whitespace is ugly and hard to read and (2) even though it is technically valid, sometimes the poor rendering of whitespace means that the Markdown renderer reads some overly-indented lines as a code block when they should not be.

To illustrate point 1, that the cleaning is so that the resulting HTML is maximally nice looking for the viewer:

|  |  |
| :- | :- |
| _Before cleaning:_ <br/> <img width="100%" src="https://user-images.githubusercontent.com/6979755/63985767-f6b48e80-ca85-11e9-8c64-5cfc72348284.png" /> | _After cleaning:_ <br/> <img width="100%" src="https://user-images.githubusercontent.com/6979755/63985963-bc97bc80-ca86-11e9-9ec3-5fb91cccc90d.png" /> |

The delta is easier to see with an external editor... Here's the same file opened in VSCode:

_Before cleaning:_ 
<img width="80%" src="https://user-images.githubusercontent.com/6979755/63985816-32e7ef00-ca86-11e9-9a56-0aa9876327cf.png" />

_After cleaning:_
<img width="80%" src="https://user-images.githubusercontent.com/6979755/63985843-4f842700-ca86-11e9-8c4e-1aebdcd00b4c.png" />

This isn't even a particularly degenerate case! Often, the resulting valid HMTL pre-cleaning is absolutely nasty.

laurent22 commented:
Thanks for clarifying and indeed it makes sense, so we can keep this html cleaning module.

In a recent commit, I've also added [html-minifier](https://www.npmjs.com/package/html-minifier), mostly to reduce the size of imported clipped pages. Do you think you can use this one to clean the Evernote notes? I guess the purpose of these modules is a bit different though, so if not possible, it's fine to keep the clean-html module.

Also when creating these notes, I think it would be better to set the `markup_language` property to `MARKUP_LANGUAGE_HTML`. This will tell Joplin that the content is only HTML (no Markdown) and make the rendering faster and more accurate. For example, it won't wrongly interpret indented HTML as a code block. This has been done for example when importing clipped content: https://github.com/laurent22/joplin/blob/ccfc80ad04d31ba3cc8b56fc0e43d65344965729/ReactNativeClient/lib/services/rest/Api.js#L470

devonzuegel commented:
> In a recent commit, I've also added [html-minifier](https://www.npmjs.com/package/html-minifier), mostly to reduce the size of imported clipped pages. Do you think you can use this one to clean the Evernote notes? I guess the purpose of these modules is a bit different though, so if not possible, it's fine to keep the clean-html module.

From a look through the docs, the minifier seems only to collapse the HTML, rather than displaying it as a nicely-indented, pretty-formatted output. Unless I'm missing some function that reverse-minifies, in which case I'm happy to use it!

> Also when creating these notes, I think it would be better to set the `markup_language` property to `MARKUP_LANGUAGE_HTML`. This will tell Joplin that the content is only HTML (no Markdown) and make the rendering faster and more accurate. For example, it won't wrongly interpret indented HTML as a code block. This has been done for example when importing clipped content:

Oh that sounds great! Can you point me to how to do it? Since the importer just receives the final string, I'm not sure where to set he `markup_language` property (or should I be setting the `convert_to` property?). Is it somewhere in [`InteropService.js`](https://github.com/devonzuegel/joplin/blob/9b1972ddeed16de607926d599778d8ad3039b00f/ReactNativeClient/lib/services/InteropService.js#L53)?

laurent22 commented:
> From a look through the docs, the minifier seems only to collapse the HTML, rather than displaying it as a nicely-indented, pretty-formatted output. Unless I'm missing some function that reverse-minifies, in which case I'm happy to use it!

Right I wasn't sure, but indeed they don't really do the same thing, so let's keep the clean-html module you've added then.

> Oh that sounds great! Can you point me to how to do it? Since the importer just receives the final string, I'm not sure where to set he markup_language property (or should I be setting the convert_to property?). Is it somewhere in InteropService.js?

`markup_language` is a property of a note, so you'd need to set it where they are created, in importEnex. The way I'd do it is for example here: https://github.com/laurent22/joplin/pull/1795/files#diff-18eba34ac5ac6744001f6fe640eb05edL218 As you've done for the note body, you could check if `importOptions.outputFormat === 'html'` and if it is, set markup_language to `MARKUP_LANGUAGE_HTML`.

Also is it possible to add a comment to explain what `newModuleFromPath_` does and why it is needed? Also the other function I wasn't sure about was `attachmentElement` - if you could add a comment next to it too that would be great.

devonzuegel commented:
Sweet, I've incorporated those two pieces of feedback. Thanks for the help!

Please do check that I handled `MARKUP_LANGUAGE_HTML` as you intended, and lmk if there's anything else before this is ready to merge. 😊

laurent22 commented:
I've tested and it's all looking good. HTML and resources are imported fine and the HTML mode also seems to be working.

There's just the package-lock conflict that needs to be fixed. You could just delete the file actually and recreate it by running `npm i` on the CliClient dir. Once this is fixed we're good to merge.

devonzuegel commented:
Hm when I do that, it doesn't seem to fix the conflict:

![image](https://user-images.githubusercontent.com/6979755/65062101-2c83bf00-d930-11e9-9737-31e2dfe77a87.png)

I've gone in and resolved the conflict by hand, but lmk if that looks wrong.

laurent22 commented:
I'm going to merge and to be sure I'll rebuilt package-lock once it's on master.

In any case thanks a lot for your great work @devonzuegel!

devonzuegel commented:
What is the process for cutting a new build for the desktop Electron app? (Again, happy to add some docs about this if helpful, once I know how it works!)

laurent22 commented:
The step to build a release is on the Travis file: https://github.com/laurent22/joplin/blob/master/.travis.yml#L55 Normally if you run these commands you should get the macOS package in the "dist" folder.

laurent22 commented:
@devonzuegel, unfortunately I had to temporarily revert this PR due to a few issues. I tried to fix some of them yesterday and I thought I was done, but today I found something else and it's harder to fix. I don't want this to block the release of the app, so that's why I'm taking it out but happy to merge it back again once it's fixed.

The first issue was that the app normally relies on auto-detection of the format based on the extension, but this capability was removed with the addition of `newModuleByFormat_`. The CLI app import was broken because of this, as well as the test units. I've tried to fix this here: 558b6443bc0799e3397a4c89ab92d387d87e26b0

The second issue is a slightly bigger one and I haven't tested everything but in particular the feature to import MD directory is broken on the desktop app (and probably CLI). This is also due to `newModuleByFormat_` which sets the module metadata to the function options, and not to the metadata. In particular, the fileExtensions array is missing and probably other properties. You've actually spotted this issue as you've added a TODO! Basically setMetdadata needs to be given the metadata defined in the `importModules` or `exportModules` array. It needs to come directly from that array, so that we can add more properties if needed.

I feel the simplest way would be to go back to using the `newModule_()` function so that we preserve the capability to auto-detect the module based on the extension. And you can add an option to disambiguate when an extension matches more than one module. That's a bit what I've started doing in my commit.

This commit reverts my changes: c7c57ab2a5681b3e09863e8b35795d7b5f7cacc9 and this one reverts yours: 50b66cceca32540eb3c06656a1918e606cd09906

Sorry about that, I wish I had spotted all this before. I guess one issue is that we don't have any unit testing for the front-end so it's easy to miss these issues.

devonzuegel commented:
Ah I'm sorry about that. Will keep an eye on that for the future, I didn't even consider that the CLI app depends on the same libraries.

> The first issue was that the app normally relies on auto-detection of the format based on the extension, but this capability was removed with the addition of `newModuleByFormat_`. The CLI app import was broken because of this, as well as the test units. I've tried to fix this here: [558b644](https://github.com/laurent22/joplin/commit/558b6443bc0799e3397a4c89ab92d387d87e26b0)

Great, thanks. Lmk if there's anything I can do to help here.

> The second issue is a slightly bigger one and I haven't tested everything but in particular the feature to import MD directory is broken on the desktop app (and probably CLI).

I'll look into this issue tonight. Thanks for finding it!

devonzuegel commented:
Okay, I believe I fixed it in this commit: https://github.com/laurent22/joplin/commit/a262839e1d268f82fbfb04452daf8c947032c82f

Digging back into this code reminded me I hadn't yet left the comment [discussed earlier](https://github.com/laurent22/joplin/pull/1795#discussion_r322379121), so I added that in this other commit, too: https://github.com/laurent22/joplin/commit/64f86ad72a20820896fcba27b928e09cec9b4627

laurent22 commented:
>  I didn't even consider that the CLI app depends on the same libraries.

I'll add this to contribution.md. It's true that it might not be obvious a lot of code is shared between all the platforms.

> Okay, I believe I fixed it in this commit: a262839

Thanks for the quick update! Please could you create a pull request with these changes (that I guess would include a revert of my reverts)?

Actually I don't quite get where you made these two commits? They seem to be on the Joplin repository but not on a branch. Is that some kind of GitHub super-power? :)

devonzuegel commented:
> Thanks for the quick update! Please could you create a pull request with these changes (that I guess would include a revert of my reverts)?

Here ya go: https://github.com/laurent22/joplin/pull/1887 😊
